### PR TITLE
[WIP]tools: add throttle mechanism to immutable object cache

### PR DESCRIPTION
**Pull request by shangdehao1**
hi, jason. @dillaman 

as we discuss,  based on existing TokenBucketThrottle class, i have implemented throttle mechanism to control promoting speed. 

many thanks.
 Signed-off-by yuan.zhou@intel.com
Signed-off-by dehao.shang@intel.com

### CODE REVIEW

File: src/tools/immutable_object_cache/ObjectCacheStore.cc [hunk 1]
```diff
@@ -35,12 +54,22 @@ ObjectCacheStore::ObjectCacheStore(CephContext *cct)
...
+    apply_qos();
+  }
+
   m_policy = new SimplePolicy(m_cct, cache_max_size, max_inflight_ops,
                               cache_watermark);
 }
 
 ObjectCacheStore::~ObjectCacheStore() {
   delete m_policy;
+  if (m_qos_enabled_flag) {
```
Comment (dillaman) [inline src/tools/immutable_object_cache/ObjectCacheStore.cc]: Nit: unneeded

File: src/tools/immutable_object_cache/ObjectCacheStore.cc [hunk 2]
```diff
@@ -97,6 +126,13 @@ int ObjectCacheStore::do_promote(std::string pool_nspace,
                    << " from pool id: " << pool_id
                    << " namespace: " << pool_nspace
                    << " snapshot: " << snap_id << dendl;
+  // try to take one token from TokenBucketThrottle
+  //   if success, directly return.
+  //   if fails, block here until new token be inserted bucket.
+  // TODO : because RO daemon just have one thread, it will block here..
```
Comment (dillaman) [inline src/tools/immutable_object_cache/ObjectCacheStore.cc]: Let's not block here -- if we don't have any budget to promote, just return `OBJ_CACHE_SKIP`

File: src/tools/immutable_object_cache/ObjectCacheStore.h [hunk 3]
```diff
@@ -47,13 +49,20 @@ class ObjectCacheStore {
...
+
   CephContext *m_cct;
   RadosRef m_rados;
   std::map<uint64_t, librados::IoCtx> m_ioctx_map;
   ceph::mutex m_ioctx_map_lock =
     ceph::make_mutex("ceph::cache::ObjectCacheStore::m_ioctx_map_lock");
   Policy* m_policy;
   std::string m_cache_root_dir;
+  bool m_qos_enabled_flag{false};
+  TokenBucketThrottle* m_throttle;
```
Comment (dillaman) [inline src/tools/immutable_object_cache/ObjectCacheStore.h]:
```suggestion
  TokenBucketThrottle* m_throttle = nullptr;
```

File: src/tools/immutable_object_cache/ObjectCacheStore.h [hunk 4]
```diff
@@ -47,13 +49,20 @@ class ObjectCacheStore {
...
+  void handle_throttle_ready(int r, C_SaferCond* cond, uint64_t f);
+
   CephContext *m_cct;
   RadosRef m_rados;
   std::map<uint64_t, librados::IoCtx> m_ioctx_map;
   ceph::mutex m_ioctx_map_lock =
     ceph::make_mutex("ceph::cache::ObjectCacheStore::m_ioctx_map_lock");
   Policy* m_policy;
   std::string m_cache_root_dir;
+  bool m_qos_enabled_flag{false};
```
Comment (dillaman) [inline src/tools/immutable_object_cache/ObjectCacheStore.h]: Nit: technically unnecessary if you just use the existence of the throttle as proof that it's enabled.

**Comment by shangdehao1** [comment]
> It would be good to support a byte throttle as well -- and test cases.

yes, i also think byte throttle is more better. currently, i am still confused about how many token for one promote request need to take. 
Because RO daemon don't know how many bytes one object contain, for example 0 bytes or 4M, how many token we need to take from throttle ?  My initial idea is that every promote request try to take 4M bytes?  how do you think this ? many thanks.

**Comment by dillaman** [comment]
> My initial idea is that every promote request try to take 4M bytes? how do you think this ? many thanks.

Pass the object size w/ the read request from the librbd client.

File: src/tools/immutable_object_cache/ObjectCacheStore.cc (hunk 1 above)
Comment (shangdehao1) [reply src/tools/immutable_object_cache/ObjectCacheStore.cc]: done. thanks jason

File: src/tools/immutable_object_cache/ObjectCacheStore.cc (hunk 2 above)
Comment (shangdehao1) [reply src/tools/immutable_object_cache/ObjectCacheStore.cc]: done. thanks

File: src/tools/immutable_object_cache/ObjectCacheStore.h (hunk 3 above)
Comment (shangdehao1) [reply src/tools/immutable_object_cache/ObjectCacheStore.h]: done. thanks

File: src/tools/immutable_object_cache/ObjectCacheStore.h (hunk 4 above)
Comment (shangdehao1) [reply src/tools/immutable_object_cache/ObjectCacheStore.h]: done. thanks

**Comment by shangdehao1** [comment]
> > My initial idea is that every promote request try to take 4M bytes? how do you think this ? many thanks.
> 
> Pass the object size w/ the read request from the librbd client.

i have modified librbd client side to pass object_len to RO daemon. many thanks for your suggestions.

File: src/common/options.cc [hunk 5]
```diff
@@ -7767,6 +7767,10 @@ static std::vector<Option> get_immutable_object_cache_options() {
     .set_default(false)
     .set_description("immutable object cache promote throttle"),
 
+    Option("immutable_object_cache_throttle_type", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
```
Comment (dillaman) [inline src/common/options.cc]: This shouldn't be needed -- it should support both concurrently. Also -- it's an unrelated change from the commit message.

File: src/librbd/cache/ParentCacheObjectDispatch.cc [hunk 6]
```diff
@@ -115,7 +115,8 @@ bool ParentCacheObjectDispatch<I>::read(
 
   m_cache_client->lookup_object(m_image_ctx->data_ctx.get_namespace(),
                                 m_image_ctx->data_ctx.get_id(),
-                                (uint64_t)snap_id, oid, std::move(ctx));
+                                (uint64_t)snap_id, object_len, oid,
```
Comment (dillaman) [inline src/librbd/cache/ParentCacheObjectDispatch.cc]: This should be the fixed object size -- not the current IO length.

File: src/common/options.cc (hunk 5 above)
Comment (shangdehao1) [reply src/common/options.cc]: sorry, i'm still confused about this comments. when one LOOKUP request arrived, how do we decide which throttle type to use ?  
Or,  librbd hook should firstly tell RO daemon which throttle type it use on register process ?

File: src/librbd/cache/ParentCacheObjectDispatch.cc (hunk 6 above)
Comment (shangdehao1) [reply src/librbd/cache/ParentCacheObjectDispatch.cc]: so sorry. i'm still not clear what's means for current IO length. could you please give some detail explain. many thanks.

File: src/common/options.cc (hunk 5 above)
Comment (dillaman) [reply src/common/options.cc]: If the user has specified a IOPS limit only -- use it. If the user has specified a bytes limit only -- use it. If a user has specified both -- use both. We do this in the IO path for librbd via multiple throttles (one for each). You always attempt to take from both throttles but only proceed once both throttles have credits available.

File: src/librbd/cache/ParentCacheObjectDispatch.cc (hunk 6 above)
Comment (dillaman) [reply src/librbd/cache/ParentCacheObjectDispatch.cc]: RBD had 4MiB default objects sizes. If I request a 1 byte IO, we still might need to promote a 4MiB object -- so the throttle should be based on the object size, not the IO request size.

Comment (shangdehao1) [reply src/librbd/cache/ParentCacheObjectDispatch.cc]: thanks jason, but how do the throttle know object size ?

Comment (dillaman) [reply src/librbd/cache/ParentCacheObjectDispatch.cc]: `m_image_ctx.get_object_size()`

File: src/common/options.cc (hunk 5 above)
Comment (shangdehao1) [reply src/common/options.cc]: hi, jason. i have refined all codes in this pull request based on your suggestions, and also modified commit messages. Additional, could i apply for access to Redhat's Teuthology lab ?

**Comment by shangdehao1** [comment]
> It would be good to support a byte throttle as well -- and test cases.

hi, jason @dillaman i also am writing some test case of Teuthology for ReadOnly cache. could i directly use Redhat's Teuthology lab to test my test case ? 

thanks

**Comment by dillaman** [comment]
@shangdehao1 I think you are referring to the Ceph community sepia lab. Access is granted to established committers to the project. In the meantime, you should be able to run tests locally and then I can help schedule the tests in sepia once you have them running locally.

**Comment by shangdehao1** [comment]
> @shangdehao1 I think you are referring to the Ceph community sepia lab. Access is granted to established committers to the project. In the meantime, you should be able to run tests locally and then I can help schedule the tests in sepia once you have them running locally.

yes, i am referring to ceph community sepia lab, many thanks for your helps. i will firstly run my test cases locally.

**Comment by shangdehao1** [comment]
> This pull request has been automatically marked as stale because it has not had any activity for 60 days. It will be closed if no further activity occurs for another 30 days.
> If you are a maintainer or core committer, please follow-up on this pull request to identify what steps should be taken by the author to move this proposed change forward.
> If you are the author of this pull request, thank you for your proposed contribution. If you believe this change is still appropriate, please ensure that any feedback has been addressed and ask for a code review.

!!!

**Comment by shangdehao1** [comment]
> @shangdehao1 I think you are referring to the Ceph community sepia lab. Access is granted to established committers to the project. In the meantime, you should be able to run tests locally and then I can help schedule the tests in sepia once you have them running locally.

hi, jason. i have added three test cases for parent cache.

File: qa/suites/rbd/cache/workloads/fio.yaml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        #debug immutable_obj_cache: 20/20
+        immutable object cache path: /tmp/ceph-immutable-object-cache
+        immutable object cache max size: 50G
+      global:
+        auth cluster required: none
+        auth service required: none
+        auth client required : none
+- exec:
+    client.0:
+      - ceph-immutable-object-cache -b
```
Comment (dillaman) [inline qa/suites/rbd/cache/workloads/fio.yaml]: Nit: don't start the daemon directly. Have the associated task start and manage the daemon.

File: qa/suites/rbd/cache/workloads/fio.yaml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    extra_packages: ['ceph-immutable-object-cache']
+- ceph:
+    conf:
+      osd:
+        osd shutdown pgref assert: true
+      client:
+        rbd_parent_cache_enabled: true
+        #debug immutable_obj_cache: 20/20
+        immutable object cache path: /tmp/ceph-immutable-object-cache
+        immutable object cache max size: 50G
```
Comment (dillaman) [inline qa/suites/rbd/cache/workloads/fio.yaml]: Hmm -- I'm not sure any teuthology node has this amount of space just sitting around idle.

File: qa/suites/rbd/cache/workloads/fio.yaml [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+- exec:
+    client.0:
+      - ceph-immutable-object-cache -b
+- rbd_fio:
+    client.0:
+      fio-io-size: 80%
+      formats: [2]
+      features: [[layering]]
+      io-engine: rbd
+      test-clone-io: 1
```
Comment (dillaman) [inline qa/suites/rbd/cache/workloads/fio.yaml]: Nit: I'm not sure this is a helpful test case. `fio` will first write to the cloned image before it reads so essentially it should never touch the immutable object cache.

File: qa/suites/rbd/cache/workloads/fio.yaml [hunk 10]
```diff
@@ -0,0 +1,32 @@
+tasks:
+- install:
+    extra_packages: ['ceph-immutable-object-cache']
+- ceph:
+    conf:
+      osd:
+        osd shutdown pgref assert: true
```
Comment (dillaman) [inline qa/suites/rbd/cache/workloads/fio.yaml]: Nit: not sure where this came from but shouldn't be needed

File: qa/suites/rbd/cache/workloads/fio.yaml [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+      client:
+        rbd_parent_cache_enabled: true
+        #debug immutable_obj_cache: 20/20
+        immutable object cache path: /tmp/ceph-immutable-object-cache
+        immutable object cache max size: 50G
+      global:
+        auth cluster required: none
```
Comment (dillaman) [inline qa/suites/rbd/cache/workloads/fio.yaml]: Nit: ditto that these overrides shouldn't be needed

File: qa/suites/rbd/cache/workloads/parent_cache.yaml [hunk 12]
```diff
@@ -0,0 +1,24 @@
+tasks:
```
Comment (dillaman) [inline qa/suites/rbd/cache/workloads/parent_cache.yaml]: Nit: same comments in this file

File: qa/suites/rbd/cache/workloads/qemu.yaml [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+        immutable object cache path: /tmp/ceph-immutable-object-cache
+        immutable object cache max size: 50G
+      global:
+        auth cluster required: none
+        auth service required: none
+        auth client required : none
+- qemu:
+    client.0:
+      clone: true
+      test: qa/workunits/true.sh
```
Comment (dillaman) [inline qa/suites/rbd/cache/workloads/qemu.yaml]: I think it would be great if you had a workload that actually needed the "golden" image to ensure it works.

File: qa/tasks/parent_cache.py [hunk 14]
```diff
@@ -0,0 +1,563 @@
+"""
+parent_cache task
```
Comment (dillaman) [inline qa/tasks/parent_cache.py]: Nit: can we rename this task to 'immutable_object_cache' instead?

File: qa/tasks/parent_cache.py [hunk 15]
```diff
@@ -0,0 +1,563 @@
...
+from teuthology import contextutil
+from tasks import rbd
+from teuthology.orchestra import run
+from teuthology.config import config as teuth_config
+
+from util.workunit import get_refspec_after_overrides
+
+log = logging.getLogger(__name__)
+
+DEFAULT_IMAGE_URL = 'http://cloud-images.ubuntu.com/releases/bionic/release/ubuntu-18.04-server-cloudimg-amd64.img'
```
Comment (dillaman) [inline qa/tasks/parent_cache.py]: I think this task should only be concerned with configuring and running the daemon. Let other tasks like qemu just run on top of a cloned image.

File: qa/tasks/qemu.py [hunk 16]
```diff
@@ -338,6 +338,18 @@ def _teardown_nfs_mount(remote, client, service_name):
             'sudo', 'systemctl', 'start', service_name
         ])
 
+def _setup_parent_cache(remote, client):
```
Comment (dillaman) [inline qa/tasks/qemu.py]: Nit: let the `immutable_object_cache` task be responsible

File: qa/suites/rbd/cache/workloads/fio.yaml (hunk 10 above)
Comment (shangdehao1) [reply qa/suites/rbd/cache/workloads/fio.yaml]: done.

File: qa/suites/rbd/cache/workloads/fio.yaml (hunk 11 above)
Comment (shangdehao1) [reply qa/suites/rbd/cache/workloads/fio.yaml]: done

File: qa/suites/rbd/cache/workloads/parent_cache.yaml (hunk 12 above)
Comment (shangdehao1) [reply qa/suites/rbd/cache/workloads/parent_cache.yaml]: done

File: qa/tasks/parent_cache.py (hunk 14 above)
Comment (shangdehao1) [reply qa/tasks/parent_cache.py]: done

File: qa/tasks/parent_cache.py (hunk 15 above)
Comment (shangdehao1) [reply qa/tasks/parent_cache.py]: hi, jason, i have deleted original parent_cache.py file, and add new file called immutable_object_cache.py. Based on qemu.py function, this new file just be responsible for setup daemon, teardown daemon and thrashes daemon. 
thanks

File: qa/tasks/qemu.py (hunk 16 above)
Comment (shangdehao1) [reply qa/tasks/qemu.py]: done

**Comment by shangdehao1** [comment]
hi, jason @dillaman I have followed your suggestions to modify test cases. 
- For fio test case
   - I use parent_cache_fio to wrapper rbd_fio task, and reuse code implemnt. 
   - In order to simulate the scenario that there are real shared objects between parent-image and child-image,  I added `thick-provision`  to at creating RBD image in rbd_fio.py file. 

- For immutable_object_cache test case 
  - I use immutable_object_cache to wrapper qemu task, and reuse code implement. And the function from qemu task will be responsible for setup qemu environment, and immutable_object_cache will be responsible to setup/teardown/thrashes daemon.

thanks

**Comment by shangdehao1** [comment]
@dillaman hi, jason. If possible, would you please help review current updates. many thanks

File: qa/suites/rbd/cache/clusters/fix-3-with-two-client.yaml [hunk 17]
```diff
@@ -0,0 +1,8 @@
+roles:
+- [mon.a, mon.c, mgr.x, osd.0, osd.1, osd.2, osd.3]
+- [mon.b, mgr.y, osd.4, osd.5, osd.6, osd.7]
+- [client.0, client.1]
+openstack:
```
Comment (dillaman) [inline qa/suites/rbd/cache/clusters/fix-3-with-two-client.yaml:5]: Nit: this is covered by the `openstack.yaml` file below

File: qa/tasks/immutable_object_cache.py [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+    based on implement of qemu task. 
+    This is task for testing immutable_object_cache.
+    """
+    assert isinstance(config, dict), \
+           "task qemu only supports a dictionary for configuration"
+
+    config = teuthology.replace_all_with_clients(ctx.cluster, config)
+
+    managers = []
+    create_images(ctx=ctx, config=config, managers=managers)
```
Comment (dillaman) [inline qa/tasks/immutable_object_cache.py:137]: I don't understand why this have any QEMU-related functionality. You should use this task just to setup and manage the lifecycle of the `ceph-immutable-cache-daemon` and use the QEMU task to launch workloads.

File: qa/tasks/qemu.py [hunk 19]
```diff
@@ -373,6 +372,7 @@ def run_qemu(ctx, config):
         if remote.os.package_type == "rpm":
             qemu_cmd = "/usr/libexec/qemu-kvm"
         args=[
+            'sudo',
```
Comment (dillaman) [inline qa/tasks/qemu.py:375]: Nit: why was this change required?

File: src/tools/immutable_object_cache/ObjectCacheStore.cc [hunk 20]
```diff
@@ -35,12 +54,39 @@ ObjectCacheStore::ObjectCacheStore(CephContext *cct)
...
+    ceph_assert(m_throttles[ROC_QOS_IOPS_THROTTLE] != nullptr);
+    delete m_throttles[ROC_QOS_IOPS_THROTTLE];
+  }
+  if (m_qos_enabled_flag & ROC_QOS_BPS_THROTTLE) {
+    ceph_assert(m_throttles[ROC_QOS_BPS_THROTTLE] != nullptr);
+    delete m_throttles[ROC_QOS_BPS_THROTTLE];
+  }
+
+  if (m_throttles[ROC_QOS_BPS_THROTTLE] != nullptr) {
+    delete m_throttles[ROC_QOS_BPS_THROTTLE];
```
Comment (dillaman) [inline src/tools/immutable_object_cache/ObjectCacheStore.cc:88]: Nit: potentially deletes it twice?

File: src/tools/immutable_object_cache/ObjectCacheStore.cc [hunk 21]
```diff
@@ -293,5 +344,73 @@ std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name,
...
+}
+
+static std::map<uint64_t, std::string> throttles_flag = {
+  { ROC_QOS_IOPS_THROTTLE, "roc_qos_iops_throttle" },
+  { ROC_QOS_BPS_THROTTLE, "roc_qos_bps_throttle" }
+};
+
+void ObjectCacheStore::apply_qos_tick_and_limit(const uint64_t flag,
+ uint64_t min_tick, uint64_t limit, uint64_t burst) {
+  static SafeTimerSingleton* safe_timer_singleton = nullptr;
```
Comment (dillaman) [inline src/tools/immutable_object_cache/ObjectCacheStore.cc:386]: Nit: unnecessary need to store static here since it will be looked up below

File: src/tools/immutable_object_cache/ObjectCacheStore.cc [hunk 22]
```diff
@@ -293,5 +344,73 @@ std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name,
...
+  }
+
+  if (got_token && (m_qos_enabled_flag & ROC_QOS_IOPS_THROTTLE)) {
+    got_token = 
+      m_throttles[ROC_QOS_IOPS_THROTTLE]->get<ObjectCacheStore, void,
+       &ObjectCacheStore::handle_throttle_ready>(object_size, this, nullptr, 0);
+  } 
+
+  if (got_token && (m_qos_enabled_flag & ROC_QOS_BPS_THROTTLE)) {
+    got_token =
```
Comment (dillaman) [inline src/tools/immutable_object_cache/ObjectCacheStore.cc:366]: What if both throttles are enabled and you are blocked by the first one? It seems like this throttle won't track BPS.

File: src/tools/immutable_object_cache/ObjectCacheStore.h [hunk 23]
```diff
@@ -30,30 +32,46 @@ class ObjectCacheStore {
...
-                    std::string object_name,
+                    uint64_t object_size, std::string object_name,
                     std::string& target_cache_file_path);
 
  private:
+  enum ThrottleTypeCode {
+    THROTTLE_CODE_BYTE,
+    THROTTLE_CODE_OBJECT
+  };
+  #define ROC_QOS_IOPS_THROTTLE 1 << 0
```
Comment (dillaman) [inline src/tools/immutable_object_cache/ObjectCacheStore.h:43]: Nit: avoid preprocessor definitions -- use an enum instead (like the one above).

File: src/tools/immutable_object_cache/Types.cc [hunk 24]
```diff
@@ -78,6 +79,7 @@ void ObjectCacheReadData::encode_payload() {
   ceph::encode(read_len, payload);
   ceph::encode(pool_id, payload);
   ceph::encode(snap_id, payload);
+  ceph::encode(object_size, payload);
```
Comment (dillaman) [inline src/tools/immutable_object_cache/Types.cc:82]: You cannot inject new fields in the middle and maintain backwards compatibility. Plus, you also need to increment the version.

File: qa/tasks/parent_cache_fio.py [hunk 25]
```diff
@@ -0,0 +1,75 @@
+"""
```
Comment (dillaman) [inline qa/tasks/parent_cache_fio.py:1]: How is this different from the `immutable_object_cache` task above? Why do we need two?
